lib/apipie_bindings/authenticators/negotiate.rb

```diff
@@ -0,0 +1,66 @@
+require 'apipie_bindings/authenticators/base'
+
+module ApipieBindings
+  module Authenticators
+    # Negotiate authenticator
+    # Implements gssapi negotiation with preexisting kerberos ticket
+    # Requires a authentication url, the authentication request will be against.
+    # This url needs to support auth negotiation and after successful auth it should return 'set-cookie' header with session.
+    # This session will be initiated in the auth request and the original request will be made with this cookie.
+    # Next requests should be already skip the negotiation, please implement Session support in your client, for not using the negotiation on every request.
+    class Negotiate < Base
+
+      # Creates new authenticator for Negotiate auth
+      # @param [String] url to make authentication request to.
+      # @param [Hash] auth_request_options passed to RestClient::Request - especially for SSL options
+      #   see https://github.com/rest-client/rest-client/blob/master/lib/restclient/request.rb.
+      # @option service service principal used for gssapi tickets - defaults to HTTP.
+      # @option method http method used for the auth request - defaults to 'get'.
+      def initialize(authorization_url, auth_request_options = {})
+        @authorization_url = authorization_url
+        @service = auth_request_options.delete(:service) || 'HTTP'
+        auth_request_options[:method] ||= 'get'
+        @auth_request_options = auth_request_options
+      end
+
+      def error(ex)
+        if ex.is_a?(GSSAPI::GssApiError)
```

Comment:
This is more of an issue with Foreman, but if the user has no permissions, they may receive an unexpected 403, which then leads to an error being raised here

```
# hammer -v -d host list
[DEBUG 2022-05-05T11:11:52 API] Using authenticator: HammerCLIForeman::Api::SessionAuthenticatorWrapper
[ERROR 2022-05-05T11:11:52 API] 403 Forbidden
[DEBUG 2022-05-05T11:11:52 API] {
    "error" => {
                    "message" => "Access denied",
                    "details" => "Missing one of the required permissions: view_hosts",
        "missing_permissions" => [
            [0] "view_hosts"
        ]
    }
}
[DEBUG 2022-05-05T11:11:52 Exception] Using exception handler HammerCLIForeman::ExceptionHandler#handle_general_exception
[ERROR 2022-05-05T11:11:52 Exception] Error: uninitialized constant ApipieBindings::Authenticators::Negotiate::GSSAPI
Error: uninitialized constant ApipieBindings::Authenticators::Negotiate::GSSAPI
[ERROR 2022-05-05T11:11:52 Exception]

NameError (uninitialized constant ApipieBindings::Authenticators::Negotiate::GSSAPI):
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/authenticators/negotiate.rb:27:in `error'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/api/negotiate_auth.rb:25:in `error'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/api/session_authenticator_wrapper.rb:66:in `error'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:248:in `rescue in http_call'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:238:in `http_call'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:190:in `call_action'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/api.rb:185:in `call'
    /usr/share/gems/gems/apipie-bindings-0.4.0/lib/apipie_bindings/resource.rb:21:in `call'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/apipie/command.rb:53:in `send_request'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/commands.rb:180:in `send_request'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/commands.rb:257:in `send_request'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/commands.rb:309:in `block in retrieve_all'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/commands.rb:308:in `loop'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/commands.rb:308:in `retrieve_all'
    /usr/share/gems/gems/hammer_cli_foreman-3.3.0.pre.develop/lib/hammer_cli_foreman/commands.rb:276:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
    /usr/share/gems/gems/hammer_cli-3.3.0.pre.develop/bin/hammer:147:in `<top (required)>'
    /usr/bin/hammer:23:in `load'
    /usr/bin/hammer:23:in `<main>'
```

```
# production.log
2022-05-05T11:11:52 [I|app|6fc2fdfe] Started GET "/api/hosts?page=1&per_page=1000" for 192.168.122.247 at 2022-05-05 11:11:52 +0000
2022-05-05T11:11:52 [I|app|6fc2fdfe] Processing by Api::V2::HostsController#index as JSON
2022-05-05T11:11:52 [I|app|6fc2fdfe]   Parameters: {"page"=>"1", "per_page"=>"1000", "apiv"=>"v2", "host"=>{}}
2022-05-05T11:11:52 [I|app|6fc2fdfe]   Rendering api/v2/errors/access_denied.json.rabl within api/v2/layouts/error_layout
2022-05-05T11:11:52 [I|app|6fc2fdfe]   Rendered api/v2/errors/access_denied.json.rabl within api/v2/layouts/error_layout (Duration: 5.9ms | Allocations: 5396)
2022-05-05T11:11:52 [I|app|6fc2fdfe] Filter chain halted as :authorize rendered or redirected
2022-05-05T11:11:52 [I|app|6fc2fdfe] Completed 403 Forbidden in 25ms (Views: 19.2ms | ActiveRecord: 1.4ms | Allocations: 13671)
```

This could possible be addressed in multiple places, attaching here as this is where the stacktrace points.